### ISSUE TITLE
Only last simple assignment gets decompiled in `x => {x.Val1 = 1; x.Val2 = 2;}`

### ISSUE THREAD

**Issue by mycroes** [hazzik/DelegateDecompiler#122]
This is somewhat of a followup to #117. Support for simple assignments was added to 0.24.0, unfortunately it only works for the last assignment. The following sample results in the decompiled expression `d => (d.Val2 = 2)`.

```csharp
public class AssignmentTest
{
    public static void Test()
    {
        Set<Data>(d =>
        {
            d.Val1 = 1;
            d.Val2 = 2;
        });
    }

    public class Data
    {
        public int Val1 { get; set; }
        public int Val2 { get; set; }
    }

    public static void Set<T>(Action<T> setter)
    {
        var d = setter.Decompile();
    }
}
```

I would expect this to generate a `BlockExpression` with two assignments, not sure if that's anything already in DelegateDecompiler that supports this. Would be willing to dive into this issue myself.

**Comment by magicmoux** [comment]
Hi @mycroes,

may I ask you what you want to achieve here, and what do you expect or need DelegateDecompiler for ?(not meaning to be offensive)

To get to the point, I just stumbled across DD to work around  third-parties' caveats (not naming EF) 
and I think I've found some outstanding ground here  (all my POCs being verified, in my humble perspective, of course) because I needed to apply complex dynamic-shares/multitenant business-logic in my DbContext queries without resorting to translate all clauses into ugly lambdas - been there done that - when native .NET implementations and in-memory LinqProvider do the job.

As far as I see, DD is a great library to get as QueryDSLProvider-agnostic as possible in the (lin)**Query-ing** area but it won't ever be able to address any underlying provider's intrinsic procedural limitations.

I think that if we want DD to have any value or usefulness, we need to decide what querying using  different providers really means.

@hazzik, any opinion/insight about this ?

Best regards,
Max

**Comment by hazzik** [comment]
@magicmoux the DD was also used in (i think now dead/dormant) [KnockoutMVC](https://github.com/AndreyAkinshin/knockout-mvc). It can be useful there. But I'm not sure if it does support assignments.

**Comment by mycroes** [comment]
I want to use DelegateDecompiler to decompile state update `Action`s. It's not related to database access in any way. To be more specific, I'd want to use it as part of our emulator infrastructure where PLC state is simulated.

**Comment by magicmoux** [comment]
Since I'm lately following the somewhat same track as you were last year, I had a look back on this issue.

@mycroes, I seem to have solved the problem based on your solution but I encountered the same stack issues as you did.
To work around this problem, I simply reverted the stack.Final method to the original implementation and handled _output_ expressions in a another List populated on Nop/Break/Ret opCodes ocurrences.

So far, it doesn't break any legacy test and satisfies yours; even if I agree that it would better to use the states.Stack object directly, but that could be addressed in a later optimization.

You may take a look at it on this last commit: 
https://github.com/magicmoux/DelegateDecompiler/commit/d14c6b005d6aeb8de3d8ba102f8fedcde25199c8

Best regards,
Max.

**Comment by magicmoux** [comment]
I tested further my solution and realized it won't perform normally expected decompilations in the inner block statements when there is no DecompileExpressionVisitor involved, since the required member accesses or calls are handled directly by the processor.

I also found out that there is yet no way to handle recursive calls during decompilation process.
So the following commit should address this issue too.

Please take a look at: 
https://github.com/magicmoux/DelegateDecompiler/commit/0fc27f8e79609cc63fa869e77c99c929440b779f

Max

**Comment by magicmoux** [comment]
Seems that my first work on this work was biased as based on an instructions queue generated in debug mode : the release instructions is much different due to optimizations and other compiler artefacts. So I'm back to the starting line on this one.

However the handling of recursive calls is not bound to the processor and I think it could _(should?)_ be integrated.

**Comment by magicmoux** [comment]
Since I'm lately following the somewhat same track as you were last year, I had a look back on this issue.

@mycroes, I seem to have lately solved the problem based on your proposal (had to start from scratch).
You can review my changes at :
https://github.com/magicmoux/DelegateDecompiler/commit/f5173f8330e1869cd9ca019bce5c8651798a4b83
please tell me if any of your tests goes awry. Currently it satisfy mines/

@hazzik, tell me whether this solution acceptable so I can submit a PR

Best regards, 
Max.

**Comment by mycroes** [comment]
Hi @magicmoux,

I'm a bit time-constrained at the moment, not sure when I'll be able to test. The case I've detailed in the issue is about as complicated as it gets for me, perhaps with nested paths to properties in worst case, so if that works it should be good enough for me. Thanks for your effort.

Regards,
Michael

**Comment by magicmoux** [comment]
Hi Michael, no problem.

I had at last (since January) some free time to define my needs and to work
on this issue, so I totally understand.

I just hope the solution is fit.

Regards,
Max


Le mer. 7 août 2019 à 20:01, Michael Croes <notifications@github.com> a
écrit :

**Comment by magicmoux** [comment]
In fact a proper and complete solution will require to dig into much OpCodes cases that were unhandled by @hazzik at first (and now I totally understand why, processor code is not  my forte !!! ) 

I'll try to submit progress as soon as stable for specific use-cases.
As for now, the solution should yet fullfill your tests' suite.
Feel free to submit more complex code you'd like to decompile (with test cases of course, please)

Max, 
(still digging into the hard code to follow the white rabbit hoping I can come back from the deep... 👍)

**Comment by hazzik** [comment]
This has been released in [`0.31.0`](https://github.com/hazzik/DelegateDecompiler/releases/tag/0.31.0). Please consider [❤️ supporting ](https://github.com/sponsors/hazzik) the project.
